File: examples/verify-webhook/python/verify.py
#!/usr/bin/env python3
"""Verify DePix App webhook signatures (Python stdlib only) against the committed
test vectors. Mirrors a real receiver per signing/webhook-signature.md:

    v1 == hex(HMAC_SHA256(secret, f"{t}.{raw_body}")), constant-time compared.

    python3 verify.py

Exit code 0 = all vectors verified; 1 = a mismatch.
"""
import hashlib
import hmac
import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
VECTORS = os.path.join(HERE, "..", "..", "..", "signing", "vectors", "webhook.json")


def parse_header(header: str) -> dict:
    out = {}
    for part in header.split(","):
        k, _, v = part.partition("=")
        out[k.strip()] = v.strip()
    return out


def verify(secret: str, header: str, raw_body: str) -> bool:
    fields = parse_header(header)
    t, v1 = fields["t"], fields["v1"]
    # A real receiver also rejects a stale `t` (e.g. abs(now - t) > 300s).
    expected = hmac.new(secret.encode(), f"{t}.{raw_body}".encode(), hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, v1)


def main() -> int:
    with open(VECTORS, encoding="utf-8") as fh:
        doc = json.load(fh)
    secret = doc["test_secret"]["value"]

    failures = 0
    for v in doc["vectors"]:
        ok = verify(secret, v["header"], v["payload"])
        print(f"{'OK  ' if ok else 'FAIL'}  {v['name']}")
        failures += 0 if ok else 1

    # Negative control: a tampered body must NOT verify.
    tampered_ok = verify(secret, doc["vectors"][0]["header"], doc["vectors"][0]["payload"] + " ")
    print(f"{'FAIL' if tampered_ok else 'OK  '}  negative control (tampered body is rejected)")
    failures += 1 if tampered_ok else 0

    if failures:
        print(f"\n{failures} check(s) failed", file=sys.stderr)
        return 1
    print(f"\nAll {len(doc['vectors'])} webhook vectors verified.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: examples/verify-webhook/node/verify.mjs
// Verify DePix webhook signatures (Node, zero deps) against the committed test
// vectors. Mirrors what a real receiver does per signing/webhook-signature.md:
//   v1 == hex(HMAC_SHA256(secret, `${t}.${rawBody}`)), constant-time compared.
//
//   node verify.mjs
//
// Exit code 0 = all vectors verified; 1 = a mismatch (or a forced-tamper that
// wrongly passed).

import crypto from "node:crypto";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const here = dirname(fileURLToPath(import.meta.url));
const vectorsPath = join(here, "..", "..", "..", "signing", "vectors", "webhook.json");
const doc = JSON.parse(readFileSync(vectorsPath, "utf8"));
const secret = doc.test_secret.value;

/** Parse "t=...,v1=..." into {t, v1}. */
function parseHeader(header) {
  const out = {};
  for (const part of header.split(",")) {
    const i = part.indexOf("=");
    out[part.slice(0, i).trim()] = part.slice(i + 1).trim();
  }
  return out;
}

/** Verify one delivery exactly as a receiver should. */
function verify(secret, header, rawBody) {
  const { t, v1 } = parseHeader(header);
  // A real receiver also rejects a stale `t` here (e.g. |now - t| > 300s); the
  // vectors use fixed timestamps, so we only check the HMAC.
  const expected = crypto.createHmac("sha256", secret).update(`${t}.${rawBody}`, "utf8").digest("hex");
  const a = Buffer.from(expected, "hex");
  const b = Buffer.from(v1, "hex");
  return a.length === b.length && crypto.timingSafeEqual(a, b);
}

let failures = 0;
for (const v of doc.vectors) {
  const ok = verify(secret, v.header, v.payload);
  console.log(`${ok ? "OK  " : "FAIL"}  ${v.name}`);
  if (!ok) failures++;
}

// Negative control: a tampered body must NOT verify.
const tamperOk = verify(secret, doc.vectors[0].header, doc.vectors[0].payload + " ");
console.log(`${tamperOk ? "FAIL" : "OK  "}  negative control (tampered body is rejected)`);
if (tamperOk) failures++;

if (failures > 0) {
  console.error(`\n${failures} check(s) failed`);
  process.exit(1);
}
console.log(`\nAll ${doc.vectors.length} webhook vectors verified.`);
